DFTBPlus calculator via ASE Python Bindings

galjos opened the pull request:
Added the `dftbplus` API via `ase` as possible `qm_prog`. Closes #115 

- [x] add checking of 3OB when setting 3rd order in QMSetup
- [x] extract smaller method from setupQMMethod()
- [x] clean up tests in testQMSetupAse.cpp and .hpp (SetUp, TearDown)
- [x] add tests in testQMSetupAse.cpp
- [x] add test for setupQMMethodAseDftbPlus()
- [x] add dispersion correction
- [x] add setup warnings to stdout

galjos commented:
Currently adding tests, will be ready soon.

galjos commented:
Test for 'src/QM' should be added in the future. @97gamjak any idea?

97gamjak commented:
> Test for 'src/QM' should be added in the future. @97gamjak any idea?

it's not that easy - maybe some simple single point calculations where we setup up the simulationbox manually?

galjos commented:
> > Test for 'src/QM' should be added in the future. @97gamjak any idea?
> 
> it's not that easy - maybe some simple single point calculations where we setup up the simulationbox manually?
Let's discuss this in more detail, before we proceed here. Otherwise the test coverage will get lower and lower.

galjos commented:
@Lucky-Armin I've tested the `third_order` keyword, and it is location dependent. If someone would first set `third_order` and after that `slakos`, the `3ob` value would overwrite the `third_order` setting.

Lucky-Armin commented:
@galjos 
Hmm, strange, when I test the it, it seems to be location independent:
no matter if I put `third_order = off;` before or after `slakos = 3ob;`, third order gets turned off;
However, this is what we wanted anyway, now i just try to move it from `QMInputParser.cpp` to `QMSettings.cpp`.

Lucky-Armin commented:
The way it is now: if `third_order`  is set before of after `slakos` is irrelevant, only thing is: 
if the user chooses to use 3ob with `third_order` turned off, it is at his or her own risk and a warning is printed in the log file when setting up the QMRunner;
Although I tested various input keyword combinations for the `ase-dftbplus` `qm_prog` there are probably still some unreasonable combinations that need to be handled, plus these combinations will get more as we include more keywords

galjos commented:
@Lucky-Armin I removed `slakosType` from the `AseDftbRunner` constructor, since `QMSetting` does the job now. And, added fixed `parseSlakosPath` test, which did not include the  `slakosType::CUSTOM`.

galjos commented:
@97gamjak IMO can be merged

97gamjak commented:
I just checked the input file logic and i don't think this will work - basically the checking if thirdorder is set can't be done in the setSlakostype function. Let's consider the case where the slakos type is set before the thirdorder key in the input file, then this value of isthirdorderset will always be false although it is found later on in the input file...

this logic weather the thirdorder key is set in the input file has to be checked in the setup of the QM method IMO. Do you agree or can you confirm this with some simple test cases?

galjos commented on src/settings/qmSettings.cpp:
```diff
@@ -251,6 +298,110 @@ void QMSettings::setQMScriptFullPath(const std::string_view &script)
+        _slakosType = THREEOB;
+        _slakosPath = __SLAKOS_DIR__ + string(_slakosType) + "/skfiles/";
+
+        if (_isThirdOrderDftbSet == true && _useThirdOrderDftb == false)
+            _useThirdOrderDftb = false;
+        else
+            _useThirdOrderDftb = true;
```
This could be simplified, we only need to check if `_isThirdOrderDftbSet` was NOT set and then set `_useThirdOrderDftb` to `true`.

galjos replied:
@97gamjak you mean this function, right?

Lucky-Armin replied:
> wa

yes, you are right this logic could be simplified like so

Lucky-Armin replied:
> This could be simplified, we only need to check if `_isThirdOrderDftbSet` was NOT set and then set `_useThirdOrderDftb` to `true`.
yes, you are right this logic could be simplified like so

Lucky-Armin commented:
i updated the code accordingly, and it produces the same results as before,
thanks  @galjos

97gamjak commented on src/settings/qmSettings.cpp:
```diff
@@ -313,9 +313,7 @@ void QMSettings::setSlakosType(const std::string_view &slakos)
-            _useThirdOrderDftb = false;
-        else
+        if (_isThirdOrderDftbSet == false)
             _useThirdOrderDftb = true;
```
ah now i get how this works... nevertheless i would strongly recommend to shift the logic from setSlakosType to the qmsetup part... this might work now but as soon as more complexitiy in the combinations is introduced it will be impossible to understand this approach.

Lucky-Armin replied:
so you mean in the file `qmSetup.cpp` under the `if (qmMethod == ASEDFTBPLUS) `
or do you mean inside the `aseDftbRunner.cpp` ?

Lucky-Armin replied:
now moved to file `qmSetup.cpp`

Lucky-Armin commented on src/setup/qmSetup.cpp:
```diff
@@ -87,6 +87,10 @@ void QMSetup::setup()
  */
 void QMSetup::setupQMMethod()
 {
+    if (QMSettings::isThirdOrderDftbSet() == false &&
+        QMSettings::getQMMethod() == QMMethod::ASEDFTBPLUS)
+        QMSettings::setUseThirdOrderDftb(true);
+
     _engine.setQMRunner(QMSettings::getQMMethod());
 }
```
I moved the enabling of third_order in the case of 3OB ASE DFTB+ calculations to the `QMSetup` class.
However, I guess this is not the way you suggested it? @97gamjak

galjos replied:
I would use:
```
if (!QMSettings::isThirdOrderDftbSet() &&
     QMSettings::getQMMethod() == QMMethod::ASEDFTBPLUS)
```
`QMSettings::isThirdOrderDftbSet() == false` is redundant.

Lucky-Armin replied:
True, respective changes are made.
Nonetheless, should we even move this logic inside the `setupQMMethod()` member function?

Lucky-Armin replied:
Functionality was moved to its own ASE DFTB+ setup member function

Lucky-Armin commented on src/setup/qmSetup.cpp:
```diff
@@ -87,6 +87,10 @@ void QMSetup::setup()
 {
+    if (QMSettings::isThirdOrderDftbSet() == false &&
+        QMSettings::getQMMethod() == QMMethod::ASEDFTBPLUS)
+        QMSettings::setUseThirdOrderDftb(true);
```
Perfect, did refactoring and only apply it to 3ob; not for matsci since its 2nd order DFTB
